### ISSUE TITLE
HydratedBloc catches failures to serialize json and continues without yielding the new state

### ISSUE THREAD

**Issue by ThinkDigitalSoftware** [felangel/hydrated_bloc#41]
Just found an issue where my HydratedBloc was rebuilding, but with the old states. There were no errors, and the state was being told to yield from the bloc correctly. Turns out the issue was that the toJson call was writing a non json-serializable map I.E. {'key1: ComplexClass}
Following the logic would not have gotten me there with the debugger
I appreciate it continuing to run, but I believe the state should still be yielded from Dart, even if it can't be yielded from json
Basically, the write should fail, not the yield

**Comment by orsenkucher** [comment]
> Basically, the write should fail, not the yield
@ThinkDigitalSoftware 

`HydratedBlocDelegate` overrides `onTransition()` method and when it fails `onError()` is invoked.
By default `Bloc` swallows errors, but you can override `onError()` either in `YourBloc` or `YourBlocDelegate` to do some logging, or throwing exceptions. 

I'm not @felangel. And in my personal opinion it's better to cause whole app panic when `toJson` is doing wrong things. To make it obvious where the problem is, it's more pragmatic I think .

**Comment by felangel** [comment]
@ThinkDigitalSoftware thanks for opening an issue!

As @orsenkucher pointed out you would likely have seen the exception of you had overridden onError in your bloc but I agree the yield shouldn’t fail so if that’s happening we should definitely address it.

@orsenkucher I don’t think it makes sense to crash the whole app if toJson fails because this is just a caching mechanism and isn’t critical for the app to function. Also, toJson might seem like it’s working for some states and then in production other states might be incorrectly serialized and cause it to fail. I would guess most developers would rather have their caching stop working rather than their entire application crash. It’s obvious in my opinion when the cache isn’t working properly and you should be able to understand why using the error/stack trace from onError. Thoughts?

**Comment by orsenkucher** [comment]
@felangel we can wrap `storage.write(... json.encode())` in `try-catch` but then problem will be lost entirely🤨
```dart
void onTransition(Bloc bloc, Transition transition) {
  super.onTransition(bloc, transition);
  final state = transition.nextState;
  if (bloc is HydratedBloc) {
    final stateJson = bloc.toJson(state);
    if (stateJson != null) {
      storage.write(bloc.storageToken, json.encode(stateJson));
    }
  }
}
```

**Comment by orsenkucher** [comment]
This code from `v3.1.0` needs revision as well 
```dart
HydratedBloc() {
  final stateJson = toJson(state);
  if (stateJson != null) {
    _storage.write(storageToken, json.encode(stateJson));
  }
}
```

Something like this?
```dart
HydratedBloc() {
  final stateJson = toJson(state);
  if (stateJson != null) {
    try {
      _storage.write(storageToken, json.encode(stateJson));
    } on dynamic catch (err, trace) {
      onError(err, trace);
    }
  }
}
```

**Comment by orsenkucher** [comment]
Maybe like so?
https://github.com/orsenkucher/hydrated_bloc/commit/37b885c7f85c6b82c0ea0ffce893a9c1f1f4b87f
These are `HydratedBlocDelegate`'s problems, let's reroute them to it

**Comment by orsenkucher** [comment]
Second option is to provide onError callback to `toJson/fromJson` methods

**Comment by felangel** [comment]
@orsenkucher let me focus on reviewing the current PRs first this afternoon and then we can decide how to proceed on this issue.

**Comment by ThinkDigitalSoftware** [comment]
It ended up throwing when I had break on all errors instead of the uncaught ones, but even still, nothing showed up on the console with that option selected

**Comment by felangel** [comment]
@ThinkDigitalSoftware can you confirm that overriding onError allows you to see the exception?

**Comment by ThinkDigitalSoftware** [comment]
Ok, let me break my code again 😁. I'll report back

**Comment by ThinkDigitalSoftware** [comment]
> @ThinkDigitalSoftware can you confirm that overriding onError allows you to see the exception?

Yes, it does.

**Comment by felangel** [comment]
@ThinkDigitalSoftware this fix is available in [v3.1.0-dev.2](https://pub.dev/packages/hydrated_bloc/versions/3.1.0-dev.2) and will be included in v3.1.0 shortly.
